x/stakeibc/keeper/hooks.go

```diff
@@ -53,20 +54,56 @@ func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochN
 				}
 			}
 		}
+
+		// DELEGATE FROM DELEGATION ACCOUNT
 		delegateInterval := int64(k.GetParam(ctx, types.KeyDelegateInterval))
 		if epochNumber%delegateInterval == 0 {
+			// get Gaia LC height
 			k.ProcessDelegationStaking(ctx)
 		}
+
 		exchangeRateInterval := int64(k.GetParam(ctx, types.KeyExchangeRateInterval))
-		if epochNumber%exchangeRateInterval == 3 { // allow a few blocks from UpdateUndelegatedBal to avoid conflicts
-			// TODO(TEST-97) update only when balances, delegatedBalances and stAsset supply are results from the same block
-			k.ProcessUpdateDelegatedBalance(ctx)
-		}
-		if epochNumber%exchangeRateInterval == 6 && (epochNumber > 100) { // allow a few blocks from UpdateDelegatedBal to avoid conflicts & wait until chain has registered zones to calc exch rate
-			// TODO(TEST-97) update only when balances, delegatedBalances and stAsset supply are results from the same block
-			k.ProcessUpdateExchangeRate(ctx)
-		}
+		if epochNumber%exchangeRateInterval == 0 && (epochNumber > 100) { // allow a few blocks from UpdateUndelegatedBal to avoid conflicts
+			// GET LATEST HEIGHT
+			// TODO(NOW) wrap this into a function
+			var latestHeightGaia int64 // defaults to 0
+			// get light client's latest height
+			connectionID := "connection-0"
+			conn, found := k.IBCKeeper.ConnectionKeeper.GetConnection(ctx, connectionID)
+			if !found {
+				k.Logger(ctx).Info(fmt.Sprintf("invalid connection id, \"%s\" not found", connectionID))
+			}
+			//TODO(TEST-112) make sure to update host LCs here!
+			clientState, found := k.IBCKeeper.ClientKeeper.GetClientState(ctx, conn.ClientId)
+			if !found {
+				k.Logger(ctx).Info(fmt.Sprintf("client id \"%s\" not found for connection \"%s\"", conn.ClientId, connectionID))
+				// latestHeightGaia = 0
+			} else {
+				// TODO(TEST-119) get stAsset supply at SAME time as gaia height
+				// TODO(TEST-112) check on safety of castng uint64 to int64
+				latestHeightGaia = int64(clientState.GetLatestHeight().GetRevisionHeight())
 
+				// TODO(119) generalize to host_zones
+				// SET STASSETSUPPLY
+				hz, _ := k.GetHostZone(ctx, "GAIA")
+				currStSupply := k.bankKeeper.GetSupply(ctx, "st"+hz.HostDenom)
+				// GET MODULE ACCT BALANCE
+				addr := k.accountKeeper.GetModuleAccount(ctx, types.ModuleName).GetAddress()
+				modAcctBal := k.bankKeeper.GetBalance(ctx, addr, hz.IBCDenom)
+
+				ControllerBalancesRecord := types.ControllerBalances{
+					Index:             strconv.FormatInt(latestHeightGaia, 10),
```

Comment:
Will this work with multiple host zones? E.g. we create a ControllerBalancesRecord for gaia at height 10 and the same for osmosis at height 10

Reply:
You might need something like f"{zoneId}.{height}"